File: vue-components/src/components/solutions/day1.js
import CryptoJS from "crypto-js";
import { MinHeap } from "@datastructures-js/heap";

export class LocationComputer {
  constructor(rows, cols) {
    // Initialize heaps and matrix for Count-Min Sketch
    this.selectionHeaps = [new MinHeap(), new MinHeap()];
    this.rows = rows;
    this.cols = cols;
    this.countMatrix = Array.from({ length: rows }, () => Array(cols).fill(0));
    this.hashFunctions = [CryptoJS.MD5, CryptoJS.SHA1, CryptoJS.SHA256];
    this.total = 0;
  }

  insertNotes(input) {
    // Insert notes into heaps based on input
    input.split("\n").forEach((line) => {
      const numbers = line.split(" ").map(Number);
      if (numbers.length != 1) {
        this.selectionHeaps[0].insert(numbers[0]);
        this.selectionHeaps[1].insert(numbers[3]);
      }
    });
  }

  insertCountMinSketch(input) {
    // Insert data into Count-Min Sketch matrix
    input.split("\n").forEach((line) => {
      const numbers = line.split(" ").map(Number);
      if (numbers.length != 1) {
        this.hashFunctions.forEach((hashFunc, i) => {
          const hashVal = parseInt(
            hashFunc(String(numbers[3])).toString(CryptoJS.enc.Hex),
            16
          );
          const bucketIdx = hashVal % this.cols;
          this.countMatrix[i][bucketIdx] += 1;
        });
      }
    });
    return this;
  }

  estimateSimilarity(elem) {
    // Estimate the frequency of an element using the Count-Min Sketch matrix
    let minCount = Infinity;
    this.hashFunctions.forEach((hashFunc, i) => {
      const hashVal = parseInt(
        hashFunc(String(elem)).toString(CryptoJS.enc.Hex),
        16
      );
      const bucketIdx = hashVal % this.cols;
      minCount = Math.min(minCount, this.countMatrix[i][bucketIdx]);
    });
    return minCount;
  }

  computeMinPairDist() {
    const aNum = this.selectionHeaps[0].extractRoot();
    const bNum = this.selectionHeaps[1].extractRoot();

    return Math.abs(aNum - bNum);
  }

  findTotalDistances() {
    // Find the total distance by processing all pairs in the heaps
    while (!this.selectionHeaps[0].isEmpty()) {
      this.total += this.computeMinPairDist();
    }
    return this.total;
  }

  findTotalSimilarity(input) {
    // Calculate total similarity based on the Count-Min Sketch
    this.total = 0;
    input.split("\n").forEach((line) => {
      const numbers = line.split(" ").map(Number);
      const freq = this.estimateSimilarity(numbers[0]);
      this.total += freq * numbers[0];
    });
    return this.total;
  }
}

export function main() {
  const part1 = (input) => {
    const lc = new LocationComputer(3, 4500);
    lc.insertNotes(input);
    return lc.findTotalDistances();
  };

  const part2 = (input) =>
    Number(
      new LocationComputer(3, 4500)
        .insertCountMinSketch(input)
        .findTotalSimilarity(input)
    );
  return { part1, part2 };
}

File: trame_aoc/app/solutions/day1.py
import heapq
import hashlib

class LocationComputer:
    def __init__(self, rows, cols):
        # Initialize heaps and matrix for Count-Min Sketch
        self.selectionHeaps = [[], []]
        self.rows = rows
        self.cols = cols
        self.countMatrix = [[0] * cols for _ in range(rows)]
        self.hashFunctions = [hashlib.md5, hashlib.sha1, hashlib.sha256]
        self.total = 0

    def insertNotes(self, input):
        """Insert notes into heaps based on input."""
        for line in input['content'].splitlines():
            numbers = [int(digit) for digit in line.split()]
            heapq.heappush(self.selectionHeaps[0], numbers[0])
            heapq.heappush(self.selectionHeaps[1], numbers[1])
        return self


    def insertCountMinSketch(self, input):
        """Insert data into Count-Min Sketch matrix."""
        for line in input['content'].splitlines():
            numbers = [int(digit) for digit in line.split()]
            for i, hashFunc in enumerate(self.hashFunctions):
                hashVal = int(hashFunc(str(numbers[1]).encode()).hexdigest(), 16)
                bucketIdx = hashVal % self.cols
                self.countMatrix[i][bucketIdx] += 1
        return self

    def estimateSimilarity(self, elem):
        """Estimate the frequency of an element using the Count-Min Sketch matrix."""
        minCount = float('inf')
        for i, hashFunc in enumerate(self.hashFunctions):
            hashVal = int(hashFunc(str(elem).encode()).hexdigest(), 16)
            bucketIdx = hashVal % self.cols
            minCount = min(minCount, self.countMatrix[i][bucketIdx])
        return minCount

    def computeMinPairDist(self):
        """Compute the minimum pair distance from the heaps."""
        if not self.selectionHeaps[0] or not self.selectionHeaps[1]:
            raise ValueError("Heaps are empty; cannot compute distance.")
        aNum = heapq.heappop(self.selectionHeaps[0])
        bNum = heapq.heappop(self.selectionHeaps[1])
        return abs(aNum - bNum)

    def findTotalDistances(self):
        """Find the total distance by processing all pairs in the heaps."""
        self.total = 0
        while self.selectionHeaps[0]:
            self.total += self.computeMinPairDist()
        tot = self.total
        return tot

    def findTotalSimilarity(self, input):
        """Calculate total similarity based on the Count-Min Sketch."""
        self.total = 0
        for line in input['content'].splitlines():
            numbers = [int(digit) for digit in line.split()]
            freq = self.estimateSimilarity(numbers[0])
            self.total += freq * numbers[0]
        return self.total
